File: StoneGameII/1140.cpp
// https://leetcode.com/problems/stone-game-ii/

class Solution
{
public:
  int stoneGameII(vector<int> &piles)
  {
    int n = piles.size();
    vector<vector<int>> dp(n, vector<int>(n, 0));
    vector<int> suffixSum(n + 1, 0);
    for (int i = n - 1; i >= 0; i--)
    {
      suffixSum[i] = suffixSum[i + 1] + piles[i];
    }
    auto aliceScore = [&](int i, int M)
    {
      return suffixSum[i] - dp[i + M][max(2 * M, n)] + dp[i + M][max(2 * M, n)];
    };
    for (int i = n - 1; i >= 0; i--)
    {
      for (int M = 1; M <= n; M++)
      {
        if (i + 2 * M >= n)
        {
          dp[i][M] = suffixSum[i];
        }
        else
        {
          for (int X = 1; X <= 2 * M; X++)
          {
            dp[i][M] = max(dp[i][M], aliceScore(i, X));
          }
        }
      }
    }

    return dp[0][1];
  }
};

File: StoneGameII/1140.js
// https://leetcode.com/problems/stone-game-ii/

var stoneGameII = function (piles) {
  const n = piles.length;
  const memo = new Array(n).fill(0).map(() => new Array(n).fill(0));
  return dp(piles, memo, 0, 1);
};
var dp = function (piles, memo, i, M) {
  if (i >= piles.length) {
    return 0;
  }
  if (memo[i][M] !== 0) {
    return memo[i][M];
  }
  let stoneGameII = 0;
  for (let X = 1; X <= 2 * M; X++) {
    if (i + X > piles.length) {
      break;
    }
    const stonesAlice = piles
      .slice(i, i + X)
      .reduce((acc, val) => acc + val, 0);
    const stonesBob = dp(piles, memo, i + X, Math.max(M, X));
    stoneGameII = Math.max(stoneGameII, stonesAlice + stonesBob);
  }
  memo[i][M] = stoneGameII;
  return stoneGameII;
};
